File: Task41/app/src/main/java/edu/deakin/s600152989/sit305/task41/TaskRepository.java
package edu.deakin.s600152989.sit305.task41;

import android.app.Application;

import androidx.lifecycle.LiveData;

import java.util.List;

public class TaskRepository {

    private TaskDao taskDao;
    private LiveData<List<Task>> allTasks;

    public TaskRepository(Application application) {
        TaskDatabase db = TaskDatabase.getInstance(application);
        taskDao = db.taskDao();
        allTasks = taskDao.getAllTasks();
    }

    public LiveData<List<Task>> getAllTasks() {
        return allTasks;
    }

    public void insert(Task task) {
        TaskDatabase.databaseWriteExecutor.execute(() -> {
            try {
                taskDao.insert(task);
            } catch (Exception e) {
                throw new RuntimeException("Insert failed", e);
            }
        });
    }

    public void update(Task task) {
        TaskDatabase.databaseWriteExecutor.execute(() -> {
            try {
                taskDao.update(task);
            } catch (Exception e) {
                throw new RuntimeException("Update failed", e);
            }
        });
    }

    public void delete(Task task) {
        TaskDatabase.databaseWriteExecutor.execute(() -> {
            try {
                taskDao.delete(task);
            } catch (Exception e) {
                throw new RuntimeException("Delete failed", e);
            }
        });
    }

    public void deleteAll() {
        TaskDatabase.databaseWriteExecutor.execute(() -> {
            try {
                taskDao.deleteAll();
            } catch (Exception e) {
                throw new RuntimeException("Delete all failed", e);
            }
        });
    }
}

File: Task41/app/src/main/java/edu/deakin/s600152989/sit305/task41/TaskRepository.kt
package edu.deakin.s600152989.sit305.task41

import android.app.Application
import androidx.lifecycle.LiveData

class TaskRepository(application: Application) {
    private val taskDao: TaskDao?
    val allTasks: LiveData<List<Task?>?>?

    init {
        val db: TaskDatabase = TaskDatabase.Companion.getInstance(application)
        taskDao = db.taskDao()
        allTasks = taskDao.allTasks
    }

    fun insert(task: Task?) {
        TaskDatabase.Companion.databaseWriteExecutor.execute(Runnable {
            try {
                taskDao!!.insert(task)
            } catch (e: Exception) {
                throw RuntimeException("Insert failed", e)
            }
        })
    }

    fun update(task: Task?) {
        TaskDatabase.Companion.databaseWriteExecutor.execute(Runnable {
            try {
                taskDao!!.update(task)
            } catch (e: Exception) {
                throw RuntimeException("Update failed", e)
            }
        })
    }

    fun delete(task: Task?) {
        TaskDatabase.Companion.databaseWriteExecutor.execute(Runnable {
            try {
                taskDao!!.delete(task)
            } catch (e: Exception) {
                throw RuntimeException("Delete failed", e)
            }
        })
    }

    fun deleteAll() {
        TaskDatabase.Companion.databaseWriteExecutor.execute(Runnable {
            try {
                taskDao!!.deleteAll()
            } catch (e: Exception) {
                throw RuntimeException("Delete all failed", e)
            }
        })
    }
}